common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,928 @@
+    __builtin_assume(mask_ != 0);
+    ssize_t index = unscaled_index();
+
+    if constexpr (ByteEncoding) {
+      // Shift to scale out of the byte encoding.
+      index >>= ByteEncodingShift;
+    }
+
+    return index;
+  }
+
+  template <typename T>
+  auto index_ptr(T* pointer) -> T* {
+    CARBON_DCHECK(mask_ != 0) << "Cannot get an index from a zero mask!";
+    __builtin_assume(mask_ != 0);
+    if constexpr (!ByteEncoding) {
+      return &pointer[unscaled_index()];
+    }
+
+    ssize_t index = unscaled_index();
+
+    // Scale the index as we counted zero *bits* and not zero *bytes*.
+    // However, we can fold that scale with the size of `T` when it is a power
+    // of two or divisible by 8.
+    CARBON_DCHECK(
+        (index & ((static_cast<size_t>(1) << ByteEncodingShift) - 1)) == 0);
+    if constexpr (sizeof(T) % 8 == 0) {
+      constexpr size_t FoldedScale = sizeof(T) / 8;
+      index *= FoldedScale;
+      return reinterpret_cast<T*>(
+          &reinterpret_cast<std::byte*>(pointer)[index]);
+    } else if constexpr (llvm::isPowerOf2_64(sizeof(T))) {
+      constexpr size_t ScaleShift = llvm::CTLog2<sizeof(T)>();
+      static_assert(ScaleShift <= ByteEncodingShift,
+                    "Scaling by >=8 should be handled above!");
+      constexpr size_t FoldedShift = ByteEncodingShift - ScaleShift;
+      index >>= FoldedShift;
+      return reinterpret_cast<T*>(
+          &reinterpret_cast<std::byte*>(pointer)[index]);
+    }
+
+    // Nothing we can fold here.
+    return &pointer[index >> ByteEncodingShift];
+  }
+
+ private:
+  // When using a byte encoding, we'll need to shift any index by this amount.
+  static constexpr size_t ByteEncodingShift = 3;
+
+  auto unscaled_index() -> ssize_t {
+    if constexpr (!ByteEncoding) {
+      // Note the cast to `size_t` to force zero extending the result.
+      return static_cast<size_t>(llvm::countr_zero(mask_));
+    } else {
+      // The index is encoded in the high bit of each byte. We compute the index
+      // by counting the number of low zero bytes there are before the first
+      // byte with its high bit set. Rather that shifting the high bit to be the
+      // low bit and counting the trailing (low) zero bits directly, we instead
+      // byte-reverse the mask and count the *leading* (high) zero bits. While
+      // this may be a wash on CPUs with direct support for counting the
+      // trailing zero bits, AArch64 only supports counting the leading zero
+      // bits and requires a bit-reverse to count the trailing zero bits. Doing
+      // the byte-reverse approach essentially combines moving the high bit into
+      // the low bit and the reverse necessary for counting the zero bits.
+      //
+      // We also cast to `size_t` to clearly zero-extend the result.
+      return static_cast<size_t>(llvm::countl_zero(llvm::byteswap(mask_)));
+    }
+  }
+
+  MaskT mask_ = 0;
+};
+
+template <typename BitIndexT>
```

Comment:
Please add a comment explaining this type. Perhaps:

```suggestion
// This is like `BitIndex`, but allows iterating through all of the matches.
template <typename BitIndexT>
```